### PR TITLE
[TECH 158] TEI-PI-PSI add/update performance enhancement

### PR DESCRIPTION
**Pull request by abyot**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramInstanceStore.java [hunk 1]
```diff
@@ -202,15 +203,23 @@ private String buildProgramInstanceHql( ProgramInstanceQueryParams params )
...
     public boolean exists( String uid )
     {
-        Integer result = jdbcTemplate.queryForObject( "select count(*) from programinstance where uid=? and deleted is false", Integer.class, uid );
-        return result != null && result > 0;
+        if ( uid == null )
+        {
+            return false;
+        }
+
+        return jdbcTemplate.queryForRowSet( "select * from programinstance where uid=+ '" + uid + "' and deleted is false limit 1;" ).next();
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramInstanceStore.java]: See "uid=+ " - seems like a + too much.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramInstanceStore.java [hunk 2]
```diff
@@ -202,15 +203,23 @@ private String buildProgramInstanceHql( ProgramInstanceQueryParams params )
...
     public boolean existsIncludingDeleted( String uid )
     {
-        Integer result = jdbcTemplate.queryForObject( "select count(*) from programinstance where uid=?", Integer.class, uid );
-        return result != null && result > 0;
+        if ( uid == null )
+        {
+            return false;
+        }
+
+        return jdbcTemplate.queryForRowSet( "select * from programinstance where uid=+ '" + uid + "' limit 1;" ).next();
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramInstanceStore.java]: See "uid=+ " - seems like a + too much?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramInstanceStore.java (hunk 1 above)
Comment (abyot) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramInstanceStore.java]: yes, unnecessary - thx.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramStageInstanceStore.java [hunk 3]
```diff
@@ -111,15 +112,23 @@ public long getProgramStageInstanceCountLastUpdatedAfter( Date time )
     @Override
     public boolean exists( String uid )
     {
-        Integer result = jdbcTemplate.queryForObject( "select count(*) from programstageinstance where uid=? and deleted is false", Integer.class, uid );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramStageInstanceStore.java]: Could we add some tests here to verify that the exists methods work?

Comment (abyot) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramStageInstanceStore.java]: yes, we have that. Here https://github.com/dhis2/dhis2-core/blob/d9ffd25026643592322df196b09bf5b542eddba0/dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/program/ProgramStageInstanceStoreTest.java#L254

**Pull request merged**